libcudacxx/include/cuda/std/detail/libcxx/include/__cccl/diagnostic.h

```diff
@@ -0,0 +1,123 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __CCCL_DIAGNOSTIC_H
+#define __CCCL_DIAGNOSTIC_H
+
+#include "../__cccl/compiler.h"
+#include "../__cccl/system_header.h"
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+// Enable us to selectively silence host compiler warnings
+#define _CCCL_TOSTRING2(_STR) #_STR
+#define _CCCL_TOSTRING(_STR)  _CCCL_TOSTRING2(_STR)
+#ifdef _CCCL_COMPILER_CLANG
+#  define _CCCL_DIAG_PUSH                _Pragma("clang diagnostic push")
+#  define _CCCL_DIAG_POP                 _Pragma("clang diagnostic pop")
+#  define _CCCL_DIAG_SUPPRESS_CLANG(str) _Pragma(_CCCL_TOSTRING(clang diagnostic ignored str))
+#  define _CCCL_DIAG_SUPPRESS_GCC(str)
+#  define _CCCL_DIAG_SUPPRESS_NVHPC(str)
+#  define _CCCL_DIAG_SUPPRESS_MSVC(str)
+#elif defined(_CCCL_COMPILER_GCC)
+#  define _CCCL_DIAG_PUSH _Pragma("GCC diagnostic push")
+#  define _CCCL_DIAG_POP  _Pragma("GCC diagnostic pop")
+#  define _CCCL_DIAG_SUPPRESS_CLANG(str)
+#  define _CCCL_DIAG_SUPPRESS_GCC(str) _Pragma(_CCCL_TOSTRING(GCC diagnostic ignored str))
+#  define _CCCL_DIAG_SUPPRESS_NVHPC(str)
+#  define _CCCL_DIAG_SUPPRESS_MSVC(str)
+#elif defined(_CCCL_COMPILER_NVHPC)
+#  define _CCCL_DIAG_PUSH _Pragma("diagnostic push")
+#  define _CCCL_DIAG_POP  _Pragma("diagnostic pop")
+#  define _CCCL_DIAG_SUPPRESS_CLANG(str)
+#  define _CCCL_DIAG_SUPPRESS_GCC(str)
+#  define _CCCL_DIAG_SUPPRESS_NVHPC(str) _Pragma(_CCCL_TOSTRING(diag_suppress str))
+#  define _CCCL_DIAG_SUPPRESS_MSVC(str)
+#elif defined(_CCCL_COMPILER_MSVC)
+#  define _CCCL_DIAG_PUSH __pragma(warning(push))
+#  define _CCCL_DIAG_POP  __pragma(warning(pop))
+#  define _CCCL_DIAG_SUPPRESS_CLANG(str)
+#  define _CCCL_DIAG_SUPPRESS_GCC(str)
+#  define _CCCL_DIAG_SUPPRESS_NVHPC(str)
+#  define _CCCL_DIAG_SUPPRESS_MSVC(str) __pragma(warning(disable : str))
+#else
+#  define _CCCL_DIAG_PUSH
+#  define _CCCL_DIAG_POP
+#  define _CCCL_DIAG_SUPPRESS_CLANG(str)
+#  define _CCCL_DIAG_SUPPRESS_GCC(str)
+#  define _CCCL_DIAG_SUPPRESS_NVHPC(str)
+#  define _CCCL_DIAG_SUPPRESS_MSVC(str)
+#endif
+
+// Convenient shortcuts to silence common warnings
```

Comment:
comment/question: Given we use the system header pragma and silence all warnings from our headers, do we think we need to still silence individual warnings?

Reply:
Yes because we test without system header